akka-stream-tests/src/test/scala/akka/stream/impl/fusing/InterpreterSpec.scala

```diff
@@ -569,39 +569,64 @@ class InterpreterSpec extends AkkaSpec with GraphInterpreterSpecKit {
 
   }
 
-  private[akka] case class Doubler[T]() extends PushPullStage[T, T] {
-    var oneMore: Boolean = false
-    var lastElem: T = _
-
-    override def onPush(elem: T, ctx: Context[T]): SyncDirective = {
-      lastElem = elem
-      oneMore = true
-      ctx.push(elem)
+  private[akka] case class Doubler[T]() extends GraphStage[FlowShape[T, T]] {
+    val out: Outlet[T] = Outlet("Doubler-out")
+    val in: Inlet[T] = Inlet("Doubler-in")
+
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+      var latest: T = _
+      var oneMore: Boolean = false
```

Comment:
no need for this boolean.

if `latest` is non-null, then it should be pushed, and set to null
if `latest` is null, then set it to the current value.

Also, for performance, this GraphStageLogic can be created as `new GraphStageLogic(shape) with InHandler with OutHandler` and implement onPush and onPull directly, which also means that it can do `setHandlers(in, out, this)`.

:)